packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,81 @@
+import * as cdk from '@aws-cdk/core';
+import { TagUpdateConstraintOptions } from '../constraints';
+import { IPortfolio } from '../portfolio';
+import { IProduct } from '../product';
+import { CfnPortfolioProductAssociation, CfnResourceUpdateConstraint } from '../servicecatalog.generated';
+import { hashValues } from './util';
+import { InputValidator } from './validation';
+
+/**
+ * Simplify type for constraint map to hold the L1 constraint resources.
+ */
+type AssociationOrConstraint = CfnPortfolioProductAssociation | CfnResourceUpdateConstraint;
+
+export class AssociationManager {
+  public static associateProductWithPortfolio(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct): void {
+    const associationKey = this.getAssociationKey(scope, portfolio, product);
+    if (!this.associationMap.has(associationKey)) {
+      const association = new CfnPortfolioProductAssociation(scope, `PortfolioProductAssociation${associationKey}`, {
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+      });
+
+      const associationValue = new Map<Constraints, AssociationOrConstraint>();
+      associationValue.set(Constraints.ASSOCIATION, association);
+      this.associationMap.set(associationKey, associationValue);
+    }
+  }
+
+  public static allowTagUpdates(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct,
+    options: TagUpdateConstraintOptions): void {
```

Comment:
This way of formatting makes this a little more palatable:

```suggestion
  public static allowTagUpdates(
    scope: cdk.Resource, portfolio: IPortfolio, product: IProduct,
    options: TagUpdateConstraintOptions,
  ): void {
```